### ISSUE TITLE
convolving messes up kernels with a sum of 0 without warning

### ISSUE THREAD

**Issue by MSeifert04** [astropy/astropy#4138]
It took me quite a while to figure out why my 2D laplacian kernel [[0,1,0],[1,-4,1],[0,1,0]] always resulted in arrays only containing NaNs without any warning at all while using astropy.convolution.convolve

The problem is, that the Exception is only raised if "normalize_kernel" is set to True but the kernel is messed up since it is divided by the sum nevertheless. 

But at least the 

> if kernel_sum < 1. / MAX_NORMALIZATION and normalize_kernel:

could be replaced by

> if kernel_sum < 1. / MAX_NORMALIZATION:

just to issue the Exception that a kernel which sum is 0 cannot be used if there is no way of including support for such kernels easily

Greetings,
Michael

edit: I am quite new to astropy and I never worked with github. So I hope I put it in the right section. Furthermore I want to apologize for my english, I am not an native and am therefore always unsure about my texts.

edit2: My first idea did not work, so I corrected the issue.

**Comment by embray** [comment]
Sounds reasonable to me--looking around, there's some largely duplicate code in the generic `Kernel.normalize()`: https://github.com/astropy/astropy/blob/5d419bbc438701e3bc99819ae5a36d9de95f5d41/astropy/convolution/core.py#L115  This *does* warn on a divide-by-zero or anything else that overflows `MAX_NORMALIZATION`.  This code could probably be factored out and used in both `convolve` and `Kernel.normalize`.  

I also don't like that the current code just raises a generic exception.  Either some more specific exception or just a warning (which can be promoted to an exception as needed) might be nice, but I'm not sure.

**Comment by MSeifert04** [comment]
I just noticed that almost the same issue was posted two years ago: https://github.com/astropy/astropy/issues/1647 .

**Comment by embray** [comment]
@MSeifert04 Yes, I think that was discussing more generally how to handle a kernel that sums to zero.  I think your issue is a little more limited to just improving how that's handled as an error state (which I agree) though it would be nice to solve the problem in general.  I'm going to assign both this and #1647 to @cdeil so it isn't forgotten about.  But he should reassign it to someone else appropriate if he doesn't have time to think about it.

**Comment by cdeil** [comment]
I haven't imported `astropy.convolve` in a long time and am busy with other things, so please un-assign me.

**Comment by embray** [comment]
Ah, according to https://github.com/astropy/astropy/wiki/Sub-package-maintainers I probably should have assigned @adonath

**Comment by MSeifert04** [comment]
Should I edit the code and create a Pull request or should the assigned maintainer do this?

**Comment by MSeifert04** [comment]
Sorry for all this "Spam" I tried to understand git and think I got some things wrong in the beginning.

**Comment by adonath** [comment]
@MSeifert04 Thanks for bringing this issue up again! As I outlined in #1647, the problem cannot be easily solved. The way the `NaN` interpolation is implemented now, it doesn't work with kernels that sum to zero, because a re-normalisation of the convolved value is required. 

The workaround in your case could be to not use a Laplacian kernel, but rather use a Laplacian of Gaussian or Mexican Hat kernel (that additionally introduces some smoothing). The idea is now to 
approximate the kernel as a difference of Gaussians (https://en.wikipedia.org/wiki/Difference_of_Gaussians). So convolve the data with one Gaussian, making use of the `NaN` interpolation, do that again with a Gaussian of slightly different width and take the difference of the individual results. Could that work for your application? This could also be a test case to modify the interpolation algorithm to work with such kind of kernel in future...

**Comment by MSeifert04** [comment]
@adonath: Yes, I have been splitting the laplacian kernel into two seperate kernels at the end. But the point of my issue and (the failed) pull-request is that currently the ``convolve`` acts unpredictable:

For example an astropy-``mexicanhat2D`` kernel with a width of 5 has a sum of 5*1e-5. With ``normalize_kernel=False`` the kernel correctly convolves the image (maybe if the image contains quite small values there might be a floating-point-deviation because of the internal normalization of the kernel). But if ``normalize_kernel=True`` the same convolution raises an Exception because the sum is too close to zero.
On the other hand a user-defined kernel with a sum of 0 is just transformed into a kernel containing NaNs without warning or exception (the exception is only raised if ``normalize_kernel=False``).
So there are two sides of the issue: Sometimes the kernel is normalized (and destroyed) without warning and sometimes an exception is raised even though the kernel could be normalized (for example my last comment on #4154). But I cannot predict with any degree of certainty if a change of ``MAX_NORMALIZATION`` would not break the convolution in different cases.

**Comment by embray** [comment]
If the normalization is only required for NaN interpolation, perhaps we could soften the requirement only to that case?  For example, we add a new boolean option `nan_interpolation=True` to the convolve functions.  If the kernel sums to zero, still raise an exception / warning like we do now, but mention that if the data does not contain NaNs the user can set `nan_interpolation=False` and continue without normalization.  Update: I see now that `convolve_fft` already has such an option, but `convolve` does not.  It even allows you to set `interpolate_nan=False` and `normalize_kernel=False` and still continue: https://github.com/astropy/astropy/blob/master/astropy/convolution/convolve.py#L436  But the warning should suggest that results are not guaranteed.

I don't understand the code for this deeply enough though to know if that's a viable workaround.

**Comment by MSeifert04** [comment]
@embray: I took a look at the cython files that compute the convolved image and it seems like the normalization is done in the NaN interpolation AND in the actual convolution. So just skipping the NaN interpolation for a kernel with a sum of zero would break the convolution even (or especially) if we skipped the interpolation. (see https://github.com/astropy/astropy/blob/v1.0.x/astropy/convolution/boundary_extend.pyx#L58 this is the normalization for interpolation and https://github.com/astropy/astropy/blob/v1.0.x/astropy/convolution/boundary_extend.pyx#L79 the actual convolution is normalized too).
But looking at the code made me think that NaN interpolation should be optional skippable (this would just be an if... in the ``boundary_*`` files since the actual convolution checks again if any element is a NaN) in case the user wants to treat NaNs as missing values instead of something that can or should be interpolated. If this would make sense I could make such a PR.

But the issue about being unpredicatable remains: The exception (or a warning) should be raised independent of the ``normalize_kernel`` state (since internally it is done). The problem thereby is that the mexicanhat would be problematic (since its kernel sum is very close to zero). This could be resolved by making the ``MAX_NORMALIZATION`` a bit less restrictive (currently a kernel-sum below 0.01 issues the exception but I guess that even a kernel-sum of 1e-5 or 1e-6 could work if the to-be-convolved-array-elements are not too close to the floating point precision limit).

**Comment by eteq** [comment]
I'm going to assign myself to this as I've done some preliminary work... *but* if someone else wants to take over, please say so and I can try to hand off what I've done...

**Comment by keflavich** [comment]
@eteq keep me in the loop - I can possibly take care of this while handling my other issues

**Comment by keflavich** [comment]
@eteq - is this not a solution?  (haven't tested, just going with an intuitive change: https://github.com/keflavich/astropy/commit/88dd50eb16c35a6baf665ee348c529dbfcf8ea56)

**Comment by keflavich** [comment]
https://github.com/astropy/astropy/pull/5649 is likely related?

**Comment by eteq** [comment]
@keflavich - that's sort of the easiest step, *but* it means the `NaN` convolution silently fails.  The issue here is more about deciding what the best way is to deal with the convolution when the kernel sums to zero, rather than the actual implementation.

**Comment by eteq** [comment]
But that said, we *could* decide just to do that and add a warning along the lines of "don't trust nan convolution if you don't normalize..."

**Comment by keflavich** [comment]
ok.  Can we discuss this further perhaps this afternoon?  I think there are deeper problems with convolution than I had appreciated this morning; see https://github.com/astropy/astropy/pull/5782

**Comment by keflavich** [comment]
I think a solution to this, in the cython code, is to just set values equal to `top` independent of the value of `bot` in the various cython codes.  I'm implementing this in #5782.  So far only 1D is done...

**Comment by keflavich** [comment]
The allowed keywords to both [`convolve`](http://keflavich-astropy.readthedocs.io/en/convolution_consistency/api/astropy.convolution.convolve.html#astropy.convolution.convolve) and [`convolve_fft`](http://keflavich-astropy.readthedocs.io/en/convolution_consistency/api/astropy.convolution.convolve_fft.html#astropy.convolution.convolve_fft) have been changed significantly.  If the option `normalize_kernel` is set to `False`, the kernel will not be normalized.

There are two options for handling NaNs: `nan_treatment='interpolate'` and `nan_treatment='fill'`.  The `interpolate` option will not work with a non-normalized kernel.

@eteq and I had a long conversation about this at astropy2017.  We concluded that we do want to support non-normalized kernels even in the presence of NaNs, but the only sensible thing to do is a two-step process using [`interpolate_replace_nans`](http://keflavich-astropy.readthedocs.io/en/convolution_consistency/api/astropy.convolution.interpolate_replace_nans.html#astropy.convolution.interpolate_replace_nans)  to replace the NaN values with sensibly-interpolated values using some normalizable kernel *before* applying the non-normalized kernel.  We need an example of this in the docs!  Is the Laplacian kernel at the top of this Issue appropriate?
